### PR TITLE
SDK-644 -- Create a base event interface

### PR DESCRIPTION
**Pull request by apeterson-branch**
## Reference
SDK-644 -- Create a base event interface

## Description
The C++ SDK development cycle started with Events, and with those events you have the ability to set a number of properties that generally relate to commerce-type events.

Much of the infrastructure that built up allowed anything that was an Event or Event subclass to be placed on the queue, and it knew how to packaged itself up and indicate which endpoint it needed to go to.

The issue here is that there are things that are saying they are Events, while they have no business having commerce type attributes.

This is simply solved by creating a new base (like, "BaseEvent"), and slightly modifying the polymorphism graph correctly.

## Testing Instructions
* Unit Tests Pass
* Samples continue to run correctly

## Risk Assessment [`MEDIUM`]
This was a fairly minor refactor, and none of the samples needed to be changed.

- [x] I, the PR creator, have tested — integration, unit, or otherwise — this code.

## Reviewer Checklist (To be checked off by the reviewer only)

- [ ] JIRA Ticket is referenced in PR title.
- Correctness & Style
    - [ ] Conforms to [Style Guides](https://google.github.io/styleguide/cppguide.html)
    - [ ] Mission critical pieces are documented in code and out of code as needed.
- [ ] Unit Tests reviewed and test issue sufficiently.
- [ ] Functionality was reviewed in QA independently by another engineer on the team.

### CODE REVIEW

File: BranchSDK/src/BranchIO/Event/BaseEvent.h [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    /**
+     * @return the API Endpoint.
+     */
+    virtual Defines::APIEndpoint getAPIEndpoint() const;
+
+    /**
+     * Prepare the event package for transmission.
+     * @param packagingInfo Context for packaging events
+     * @param jsonPackage result of the packaging process
+     * @todo(andyp) Revisit Scope
```
Comment (bklastaitis-branch) [inline BranchSDK/src/BranchIO/Event/BaseEvent.h]: 👍 re revisiting scope. I'd argue `packageV1Event` and `packageV2Event` should be virtual, not `package `

File: BranchSDK/src/BranchIO/Event/BaseEvent.h [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+     * @param jsonPtr Base JSON to start with
+     */
+    BaseEvent(Defines::APIEndpoint apiEndpoint, const std::string &eventName, JSONObject::Ptr jsonPtr = nullptr);
+
+    /**
+     * Add a Raw Event Property Name and Value
+     * @param propertyName Property Name
+     * @param propertyValue Property Value
+     * @return this object for chaining builder methods
+     */
```
Comment (bklastaitis-branch) [inline BranchSDK/src/BranchIO/Event/BaseEvent.h]: I noticed that custom data is only supported in v2 endpoints but the API `addCustomDataProperty` is not restricted to to v2 events. Maybe we should mention in the documentation here (also potentially make a ticket for it).

File: BranchSDK/src/BranchIO/Event/BaseEvent.cpp [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+BaseEvent::addEventProperty(const char *propertyName, double propertyValue) {
+    addProperty(propertyName, propertyValue);
+    return *this;
+}
+
+BaseEvent&
+BaseEvent::addCustomDataProperty(const std::string &propertyName, const std::string &propertyValue) {
+
+    // These go into "custom_data"
+    if (mCustomData.isNull()) {
```
[the remark this answers is not in the document]
Comment (apeterson-branch) [reply BranchSDK/src/BranchIO/Event/BaseEvent.cpp]: A JSONObject::Ptr can be non-null, but the property it holds might be null.

[the remark this answers is not in the document]
Comment (bklastaitis-branch) [reply BranchSDK/src/BranchIO/Event/BaseEvent.cpp]: That makes a lit more sense 😅

File: BranchSDK/src/BranchIO/Event/BaseEvent.h (hunk 1 above)
Comment (bklastaitis-branch) [reply BranchSDK/src/BranchIO/Event/BaseEvent.h]: Talked in person, it makes more sense to override `package` if, for example, we are implementing some funky request that's neither v1 or v2, so we are keeping implementation as is.

File: BranchSDK/src/BranchIO/Event/BaseEvent.h (hunk 2 above)
Comment (bklastaitis-branch) [reply BranchSDK/src/BranchIO/Event/BaseEvent.h]: Note, this comment was supposed to be placed above `addCustomDataProperty`, not `addEventProperty`, so the request is to add an explanation in the description of the method `addCustomDataProperty`

Comment (apeterson-branch) [reply BranchSDK/src/BranchIO/Event/BaseEvent.h]: Added comment

**Pull request merged**